lib/model/compose.dart

```diff
@@ -0,0 +1,103 @@
+import 'dart:math';
+
+//
+// Put functions for nontrivial message-content generation in this file.
+//
+// If it's complicated enough to need tests, it should go in here.
+//
+
+// https://spec.commonmark.org/0.30/#fenced-code-blocks
+final RegExp _openingBacktickFenceRegex = (() {
+  // Recognize a fence with "up to three spaces of indentation".
+  // Servers don't recognize fences that start with spaces, as of Server 7.0:
+  //   https://chat.zulip.org/#narrow/stream/6-frontend/topic/quote-and-reply.20fence.20length/near/1588273
+  // but that's a bug, since those fences are valid in the spec.
+  // Still, it's harmless to make our own fence longer even if the server
+  // wouldn't notice the internal fence that we're steering clear of,
+  // and if servers *do* start following the spec by noticing indented internal
+  // fences, then this client behavior will be nice.
+  const lineStart = r'^ {0,3}';
+
+  // The backticks, captured so we can see how many.
+  const backticks = r'(`{3,})';
+
+  // The "info string" plus (meaningless) leading or trailing spaces or tabs.
+  // It can't contain backticks.
+  const trailing = r'[^`]*$';
+  return RegExp(lineStart + backticks + trailing, multiLine: true);
+})();
+
+/// The shortest opening backtick fence that's longer than any in [content].
```

Comment:
I think the reasoning here can actually be simplified by removing the word "opening", from this summary and the function's name. We're finding a backtick fence that's longer than *any backtick fence*, whether opening or closing, in the content.

That's what the code is actually already doing, because the regex matches closing fences as well as opening fences.

It's easier to reason about, because a fence that's an opening fence in the original content in isolation could become a closing fence, if we were to try to enclose it in fences that were too short.

And it's what we need to do:
* We need our opening fence to be long enough that it won't be closed by any fence in the content.
* We need our closing fence to be long enough that it will close any outstanding opening fences in the content.